map/gps_track.cpp

```diff
@@ -290,30 +259,146 @@ void GpsTrack::LazyInitCollection()
+
+  UpdateFile(needClear, points);
+
+  if (!m_collection)
+    return;
+
+  UpdateCollection(duration, needClear, points);
+}
+
+bool GpsTrack::HasCallback()
+{
+  lock_guard<mutex> lg(m_callbackGuard);
+  return m_callback != nullptr;
+}
+
+void GpsTrack::UpdateFile(bool needClear, vector<TItem> const & points)
+{
+  // Update file, if need
+  // If file exception happens, then drop the file.
+
+  LazyInitFile();
+
+  if (!m_file)
+    return;
+
+  try
+  {
+    // clear points from file, if need
+    if (needClear)
+      m_file->Clear();
+
+    // add points to file if need
+    for (auto const & point : points)
+    {
+      size_t evictedId;
+      m_file->Append(point, evictedId);
+    }
+  }
+  catch (RootException & e)
+  {
+    LOG(LINFO, ("GpsTrackFile.Append has caused exception:", e.Msg()));
+    m_file.reset();
+  }
+}
+
+void GpsTrack::UpdateCollection(hours duration, bool needClear, vector<TItem> const & points)
 {
-  // Must be called under m_guard lock
+  // Apply Clear, SetDuration and Add points
+
+  // Clear points from collection, if need.
+  pair<size_t, size_t> evictedIdsByClear = make_pair(kInvalidId, kInvalidId);
+  if (needClear)
+    evictedIdsByClear = m_collection->Clear(false /* resetIds */);
+
+  // Set duration for collection, if need
+  // Set duration before Add because new duration can be more than previous value.
+  pair<size_t, size_t> evictedIdsByDuration = make_pair(kInvalidId, kInvalidId);
+  if (duration != m_collection->GetDuration())
+    evictedIdsByDuration = m_collection->SetDuration(duration);
+
+  // Add points to the collection, if need
+  pair<size_t, size_t> evictedIds = make_pair(kInvalidId, kInvalidId);
+  pair<size_t, size_t> addedIds = make_pair(kInvalidId, kInvalidId);
+  if (!points.empty())
+    addedIds = m_collection->Add(points, evictedIds);
+
+  // Result evicted is
+  evictedIds = UnionRanges(evictedIds, UnionRanges(evictedIdsByClear, evictedIdsByDuration));
+
+  // Send callback notification.
+  // Callback must be protected by m_callbackGuard
+
+  lock_guard<mutex> lg(m_callbackGuard);
 
   if (!m_callback)
     return;
 
-  // Get points from collection to send them to the callback
-  vector<pair<size_t, location::GpsTrackInfo>> toAdd;
-  toAdd.reserve(m_collection->GetSize());
-  m_collection->ForEach([&toAdd](location::GpsTrackInfo const & info, size_t id)->bool
+  if (m_needSendSnapshop)
   {
-    toAdd.emplace_back(id, info);
-    return true;
-  });
+    m_needSendSnapshop = false;
 
-  if (toAdd.empty())
-    return; // nothing to send
+    // Get all points from collection to send them to the callback
+    vector<pair<size_t, TItem>> toAdd;
+    toAdd.reserve(m_collection->GetSize());
+    m_collection->ForEach([&toAdd](TItem const & point, size_t id)->bool
+    {
+      toAdd.emplace_back(id, point);
+      return true;
+    });
+
+    if (toAdd.empty())
+      return; // nothing to send
+
+    m_callback(move(toAdd), make_pair(kInvalidId, kInvalidId));
```

Comment:
it is special lock for callback which prevents SetCallback and call callback at the same time.
what if user tries detach while worker thread calls thread.